File: searching/firstBadVersion.py
# https://leetcode.com/problems/first-bad-version/
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        i = 1
        j = n
        while i < j:
            k = (i+j)//2
            if isBadVersion(k):
                j = k
            else:
                i = k+1
        return i

File: binarySearch/firstBadVersion.js
// https://leetcode.com/problems/first-bad-version/
/**
 * @param {function} isBadVersion()
 * @return {function}
 */
const solution = (isBadVersion) => {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return (n) => {
    // binary search
    let start = 1,
      end = n;
    while (start < end) {
      const mid = Math.floor((start + end) / 2);
      if (isBadVersion(mid)) {
        end = mid; // look on left side of mid
      } else {
        start = mid + 1; // look on the right side of mid
      }
    }
    return start;
  };
};
